pydantic_ai_slim/pydantic_ai/durable_exec/temporal/_dynamic_toolset.py

```diff
@@ -0,0 +1,172 @@
+    """Temporal wrapper for DynamicToolset.
+
+    This provides static activities (get_tools, call_tool) that are registered at worker start time,
+    while the actual toolset selection happens dynamically inside the activities where I/O is allowed.
+    """
+
+    def __init__(
+        self,
+        toolset: DynamicToolset[AgentDepsT],
+        *,
+        activity_name_prefix: str,
+        activity_config: ActivityConfig,
+        tool_activity_config: dict[str, ActivityConfig | Literal[False]],
+        deps_type: type[AgentDepsT],
+        run_context_type: type[TemporalRunContext[AgentDepsT]] = TemporalRunContext[AgentDepsT],
+    ):
+        super().__init__(toolset)
+        self.activity_config = activity_config
+        self.tool_activity_config = tool_activity_config
+        self.run_context_type = run_context_type
+
+        async def get_tools_activity(params: _GetToolsParams, deps: AgentDepsT) -> dict[str, _ToolInfo]:
+            """Activity that calls the dynamic function and returns tool definitions."""
+            ctx = self.run_context_type.deserialize_run_context(params.serialized_run_context, deps=deps)
+
+            async with self.wrapped:
+                tools = await self.wrapped.get_tools(ctx)
+                return {
+                    name: _ToolInfo(tool_def=tool.tool_def, max_retries=tool.max_retries)
+                    for name, tool in tools.items()
+                }
+
+        get_tools_activity.__annotations__['deps'] = deps_type
+
+        self.get_tools_activity = activity.defn(name=f'{activity_name_prefix}__dynamic_toolset__{self.id}__get_tools')(
+            get_tools_activity
+        )
+
+        async def call_tool_activity(params: CallToolParams, deps: AgentDepsT) -> CallToolResult:
+            """Activity that instantiates the dynamic toolset and calls the tool."""
+            ctx = self.run_context_type.deserialize_run_context(params.serialized_run_context, deps=deps)
+
+            async with self.wrapped:
+                tools = await self.wrapped.get_tools(ctx)
+                tool = tools.get(params.name)
+                if tool is None:  # pragma: no cover
+                    raise UserError(
+                        f'Tool {params.name!r} not found in dynamic toolset {self.id!r}. '
+                        'The dynamic toolset function may have returned a different toolset than expected.'
+                    )
+
+                return await self._call_tool_in_activity(params.name, params.tool_args, ctx, tool)
+
+        call_tool_activity.__annotations__['deps'] = deps_type
+
+        self.call_tool_activity = activity.defn(name=f'{activity_name_prefix}__dynamic_toolset__{self.id}__call_tool')(
+            call_tool_activity
+        )
+
+    @property
+    def temporal_activities(self) -> list[Callable[..., Any]]:
+        return [self.get_tools_activity, self.call_tool_activity]
+
+    async def __aenter__(self) -> Self:
```

Comment:
This behavior in aenter/aexit may make sense on `TemporalWrapperToolset`. Then we can also remove the overrides from `TemporalMCPServer`.